cpp/src/arrow/util/uri.cc

```diff
@@ -218,6 +215,9 @@ std::string Uri::path() const {
 std::string Uri::query_string() const { return TextRangeToString(impl_->uri_.query); }
 
 Result<std::vector<std::pair<std::string, std::string>>> Uri::query_items() const {
+  // XXX would it be worthwhile to fold this parsing into Uri::parse() or maybe
+  // cache these lazily in an unordered_map? Then we could provide
+  // Uri::query_item(std::string name)
```

Comment:
Not all URIs may use this format for query strings, so this can't be folded into `Parse`.

We can perhaps cache the result and also provide a `query_items_map()` to return an `unordered_map` if we want.

Reply:
For URIs which do not use this format, query_items will always return Status::Invalid. This doesn't preclude parsing them at construction time; the data member would be `optional<unordered_map<string, string>> query_item_map_;` where `nullopt` indicates that the query string didn't have key=value format.

Out of curiosity, which URIs don't use this format?

Reply:
> This doesn't preclude parsing them at construction time

Why not. Hopefully there are no cases where parsing a URI is performance-critical.

> Out of curiosity, which URIs don't use this format?

Concretely, I don't know, but we cannot preclude this :-) The RFC does not mandate any particular syntax for the query component of a URI: https://datatracker.ietf.org/doc/html/rfc3986#section-3.4